### ISSUE TITLE
Larger input image, Bigger model size

### ISSUE THREAD

**Issue by minhohihi** [tf-coreml/tf-coreml#194]
I'm facing mlmodel size issue on iOS when implementing style transfer model. tfcoreml convertor makes mlmodel bigger when setting large input image size. Does size of mlmodel depend on size of input image? I think mlmodel size should be constant despite of input image size because mlmodel only contains weights trained.

**Comment by aseemw** [comment]
Please provide details to replicate the issue.

**Comment by minhohihi** [comment]
@aseemw Sorry for late.

I made decoder like below
     Input: feature from encoder as shape as 1x120x160x512 (N x H x W x C)
     Net: 
               deconv = Input
               rows = Input.W
               cols = Input.H
               for i in range(3):
                    deconv = tf.image.resize_images(deconv, rows ,cols, 
                                                  method=tf.image.ResizeMethod.NEAREST_NEIGHBOR)
                    deconv = tf.pad(deconv, [[0, 0], [padding_size, padding_size ], 
                                                 [padding_size, padding_size ], [0, 0]], mode='REFLECT')
                    W_conv, b_conv = self.deconv_layer[i]
                    deconv = conv2d(deconv, W_conv, stride=self.dec_conv_stride[i], padding='VALID')
                    deconv = tf.nn.bias_add(deconv, b_conv)
                    rows = rows * 2
                    cols = cols * 2

Input shape of decoder can be changed by input shape of encoder. Large input image makes big feature. (ex. If input image of encoder is 360x480, then feature shape is 45x60x512. And if input image size is 720x960, then feature shape is 90x120x512). 

First I freeze above network (model_frozen.pb) then convert model_frozen.pb to model_frozen_coreml.mlmodel. As I reported before size of converted .mlmodel is dependent on size of input feature although model_frozen.pb file always has same size.

I just wonder why this happened and please let me know how to solve this problem.
Thank you~

**Comment by minhohihi** [comment]
Here is the frozen .pb

[arbitrary_enc_decoder_frozen.pb.zip](https://github.com/tf-coreml/tf-coreml/files/2092556/arbitrary_enc_decoder_frozen.pb.zip)

**Comment by minhohihi** [comment]
Here is the frozen .pb

[frozen.pb.zip](https://github.com/tf-coreml/tf-coreml/files/2092559/frozen.pb.zip)

Thank you~

**Comment by aseemw** [comment]
I figured out why this is happening. Its basically due of the limitation of the "minimum" and "maximum" layers in Core ML. They do not support broadcasting. 
The TF graph has two operations in the end: min(x,255) and max(x,0). Currently these are implemented using CoreML's load constant layer and min/max layers. Due to the load constant layer and broadcasting its shape during conversion, the size of the model increases. 

Long term solution is not to change the converter but to modify the CoreML layer. 
Short term: maybe a hack in the converter can get rid of this issue, for this particular model. max(x,0) is same as relu. min(x,255) is trickier but can be approximated using sigmoid hard layer.

If these ops are in the end, you can also convert the graph before them and implement them as post processing after obtaining the output from CoreML. 

<img width="594" alt="screen shot 2018-06-11 at 11 13 03 pm" src="https://user-images.githubusercontent.com/7546478/41273327-d17789d0-6dcd-11e8-8f94-02133bc60a96.png">

**Comment by minhohihi** [comment]
@aseemw I really appreciate your detail review. Your guide helped solving problem. Thank you :).

**Comment by daxiangpanda** [comment]
@aseemw can you please teach me how to convert such pb models using tf-coreml?I dont know how to set the input shape in the param input_name_shape_dict.this is my pb file:https://drive.google.com/file/d/1bGM82rdAQlZUj7F8W2wm_JJBmBxNBiah/view?usp=sharing
thank you!

**Comment by aseemw** [comment]
These are the two inputs (from this [script](https://github.com/tf-coreml/tf-coreml/blob/master/utils/inspect_pb.py))

```
0: op name = import/Placeholder, op type = ( Placeholder ), inputs = , outputs = import/Placeholder:0
@input shapes:
@output shapes:
name = import/Placeholder:0 : (?, ?, 3)
1: op name = import/Placeholder_1, op type = ( Placeholder ), inputs = , outputs = import/Placeholder_1:0
@input shapes:
@output shapes:
name = import/Placeholder_1:0 : (?, ?, 3)
```



So the input dict should be something like: {"Placeholder:0":[H1,W1,3], "Placeholder_1:0": [H2,W2,3]}

where H1,H2,W1,W2 are the values that you have tested works on the original TF model. 

However, looking at the ops in this graph, there are quite a few that are not supported at this time: 

<img width="423" alt="screen shot 2018-08-15 at 9 52 37 am" src="https://user-images.githubusercontent.com/7546478/44161234-8fda5800-a071-11e8-8b85-ce58b8268162.png">

**Comment by mayankb009** [comment]
Hi, I was having a similar issue when my model size increased from 30MB (.pb file) to 90MB (.mlmodel) after the conversion.
Thanks to @aseemw for pointing out the root cause of this. For a short term solution, I replaced the Minimum layer implementation in the converter as follows:-
``` 
def minimum(op, context):
  input_name = make_tensor(op.inputs[0], context)
  output_name = compat.as_str_any(op.outputs[0].name)
  neg_output_name = output_name + '_neg'
  # negate it
  context.builder.add_activation(
      neg_output_name, 'LINEAR', input_name, neg_output_name, [-1.0, 0])
  # apply threshold
  clip_output_name = output_name + '_clip'
  context.builder.add_unary(
      clip_output_name, neg_output_name, clip_output_name, 'threshold',
      alpha=-6.0)
  # negate it back
  context.builder.add_activation(
      output_name, 'LINEAR', clip_output_name, output_name, [-1.0, 0])
  context.translated[output_name] = True
```
For me the constant was 6 for all the layers, so it became fairly easy. It might be tricky if all your layers have a different constant to compare against. Doing this implementation helps me get rid of the broadcasting of constant to huge matrices which was increasing the model size. Verified that converted model is same size and produces same results.
